lib/compiler/src/compiler.rs

```diff
@@ -81,21 +81,29 @@ pub trait Compiler: Send {
         features: &Features,
         data: &'data [u8],
     ) -> Result<(), CompileError> {
-        let mut validator = Validator::new();
+        assert!(
+            !features.module_linking,
+            "Module linking feature currently not supported by wasmparser"
+        );
         let wasm_features = WasmFeatures {
             bulk_memory: features.bulk_memory,
             threads: features.threads,
             reference_types: features.reference_types,
             multi_value: features.multi_value,
             simd: features.simd,
             tail_call: features.tail_call,
-            module_linking: features.module_linking,
             multi_memory: features.multi_memory,
             memory64: features.memory64,
             exceptions: features.exceptions,
             deterministic_only: false,
+            component_model: false,
+            extended_const: false,
+            mutable_global: true,
+            relaxed_simd: false,
+            saturating_float_to_int: true,
```

Comment:
I think this _may_ be a protocol version change for us… Not a huge deal right this moment though since the pwasm parser won’t support this anyway, but probably should make sure we don’t mess this up somehow. Same for the `sign_extension` I believe.

Reply:
Oh. Welp, and say I thought these were probably already enabled before… Just to check, as I also set mutable_global to true, do we agree that this one should be left to true or should it be false too?

(Also I guess this ties back to the discussion about wasmer behavior changes that we need to have)

Reply:
`mutable_globals` were always supported and should stay true AFAIK. I believe this option was added primarily as a sanitization option (i.e. somebody probably wanted an ability to disallow mutation of the globals?)

Reply:
Another observation, these should be passed in from nearcore in the first place, rather than hardcoded in wasmer.

But also, which tests exactly?

Reply:
I’m of the exact opposite opinion: IMO, especially as we’re going towards vendoring wasmer into nearcore, wasmer should expose which features it supports and nearcore should just use the constant defined there; and we should discard from the codebase anything related to extensions we don’t have immediate plans on supporting. However, such a refactoring should probably happen only once wasmer gets merged into nearcore, to avoid complications.

As for which tests, turning `saturating_float_to_int` to false makes the following tests fail:
- `wast::spec::binary_leb128::singlepass::universal`
- `wast::spec::conversions::singlepass::universal`

And turning `sign_extensions` to false makes the following tests fail:
- `wast::spec::i64::singlepass::universal`
- `wast::spec::i32::singlepass::universal`

So I think that these flags were just implicitly turned on before (like mutable-globals), and turning them off would be a protocol change (and though anyway such a big wasmparser bump would most likely be a protocol change anyway, it’d actually be reducing our scope and so likely to break existing contracts)

Reply:
> So I think that these flags were just implicitly turned on before

Could be the case for wasmer, sure. But pwasm parser definitely does not know how to parse these instructions and so we definitely were rejecting any contracts that were using these features in the past.

> wasmer should expose which features it supports and nearcore should just use the constant defined there; and we should discard from the wasmer codebase anything related to extensions we don’t have immediate plans on supporting

The problem with this approach is that wasmer does not really have a good overview of the gcd of features supported across all of the nearcore’s components. So in practice nearcore is the only thing that can come up with the set of features that works for it (and the protocol.)

I do think it makes sense to discard anything and everything we don’t want to support right now though. Having unused code is sketchy and runs a risk of us accidentally stabilizing that's non-compliant (to the spec.) Any new instructions and wasm features that we enable should be a conscious decision and getting that requires a significant amount of scrutiny and validation.

> As for which tests

I’ll double check why they were passing, though I won’t be surprised if the older wasmparser versions didn’t reject them.

Reply:
I’m not sure what the link is between pwasm and here? Right now the only thing I’m doing here is, bump the wasmparser dependency in wasmer, trying not to change any observable behavior outside of wasmer.

What I can say for sure, based on the wasmer tests no longer passing if setting these flags to false, is that if wasmparser had had these flags in 0.7x then they would have defaulted to `true`, as the wasmer tests would not have passed otherwise. And so, that setting them to `false` here would result in an observable behavior change outside of wasmer.

As for moving the feature list into wasmer instead of nearcore, I guess this is related with my ongoing work of integrating finite-wasm into wasmer directly: once that’s done we can get rid of pwasm altogether, and then wasmer will actually become the only part where there’d actually be a supported feature list (unless I missed another one)

Reply:
Okay, right, so the feature flags for these were added in https://github.com/bytecodealliance/wasm-tools/commit/c8a5246026c9dd66b0ae7d08520a3b38b0338493, well after these operations were introduced, so it makes sense that the test suite would start failing once they are disabled.

> I’m not sure what the link is between pwasm and here?

pwasm does not support these instructions AFAIK. If the support for them is hardcoded to `true` here, we run a higher than necessary risk of enabling these instructions unintentionally when we remove pwasm-based preparation code in nearcore. In particular what worries me is somebody making a decision based on these support flags being hardcoded here to make a choice on the nearcore side. https://github.com/near/nearcore/pull/8119 for example sets the flag(s) in question to true too. Arguably that's entirely an issue with `nearcore`, but given our plans to merge wasmer into nearcore and that we would eventually start integrating the codebases further (e.g. removing a superfluous validation pass), it would be pretty easy to miss the detail here.

I think at the very least these should remain configurable, even if defaulting to enabled, but I definitely would be more comfortable with the flag disabled by default and enabled explicitly for the tests [somewhere here, perhaps?](https://github.com/near/wasmer/blob/52464a0baeffd35d6b2c40acf0d259b69157355a/tests/compilers/wast.rs#L22-L36)